FWCore/TestModules/plugins/GenericCloner.cc

```diff
@@ -0,0 +1,231 @@
+          for (auto& pattern : eventPatterns_) {
+            if (pattern.match(product)) {
+              // check that the product is not transient
+              if (product.transient()) {
+                edm::LogWarning("GenericCloner") << "Event product " << product.branchName() << " of type "
+                                                 << product.unwrappedType() << " is transient, will not be cloned.";
+                break;
+              }
+              if (verbose_) {
+                edm::LogInfo("GenericCloner")
+                    << "will clone Event product " << product.branchName() << " of type " << product.unwrappedType();
+              }
+              Entry entry;
+              entry.objectType_ = product.unwrappedType();
+              entry.wrappedType_ = product.wrappedType();
+              // TODO move this to EDConsumerBase::consumes() ?
+              entry.getToken_ = this->consumes(
+                  edm::TypeToGet{product.unwrappedTypeID(), edm::PRODUCT_TYPE},
+                  edm::InputTag{product.moduleLabel(), product.productInstanceName(), product.processName()});
+              entry.putToken_ = this->produces(product.unwrappedTypeID(), product.productInstanceName());
+              eventProducts_.emplace_back(std::move(entry));
+              break;
+            }
+          }
+          break;
+
+        case edm::InLumi:
+        case edm::InRun:
+        case edm::InProcess:
+          // lumi, run and process products are not supported
+          break;
+
+        default:
+          throw edm::Exception(edm::errors::LogicError)
+              << "Unexpected product type " << product.branchType() << "\nPlease contact a Framework developer.";
+      }
+    });
+  }
+
+  void GenericCloner::produce(edm::StreamID /*unused*/, edm::Event& event, edm::EventSetup const& /*unused*/) const {
+    for (auto& product : eventProducts_) {
+      edm::Handle<edm::WrapperBase> handle(product.objectType_.typeInfo());
+
+      event.getByToken(product.getToken_, handle);
+
+      // create a "wrapper", whose wrapped product will be cloned into "clone"
+      edm::WrapperBase const* wrapper = handle.product();
+      std::unique_ptr<edm::WrapperBase> clone(static_cast<edm::WrapperBase*>(product.wrappedType_.getClass()->New()));
+
+      // get a "serialiserSource" object from the plugin factory.
+      // serializeSource objects can produce, for a given type, both const and mutable TrivialSerialisers.
+      std::unique_ptr<ngt::TrivialSerialiserSourceBase> serialiserSource{
+          ngt::TrivialSerialiserSourceFactory::get()->create(product.objectType_.typeInfo().name())};
+
+      // initialise a const and a mutable TrivialSerialisers.
+      auto const_serialiser = serialiserSource->initialize(*wrapper);
+      auto mutable_serialiser = serialiserSource->initialize(*clone);
+
+      if (mutable_serialiser->hasTrivialCopyTraits()) {
+        // mark the clone as present
+        clone->markAsPresent();
```

Comment:
Can you move this to the non-`const` `TrivialSerialiserSource<T>::initialize(edm::WrapperBase& wrapper) { ... }` ?

That way it can be called directly on the object of type `Wrapper<T>`, and there is no need to add `markAsPresent()` to the `WrapperBase` interface.